File: 42-trapping-rain-water/trapping-rain-water.py
class Solution(object):
    def trap(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        prefixList = [0]
        prefixMax = 0

        for i in range(1, len(height)):
            prefixMax = max(prefixMax, height[i - 1])
            prefixList.append(prefixMax)

        suffixList = [0] * len(height)
        suffixMax = 0

        for i in range(len(height) - 2, -1, -1):
            suffixMax = max(suffixMax, height[i + 1])
            suffixList[i] = (suffixMax)
        trapped = 0
        for i in range(len(height)):
            if height[i] < suffixList[i] and height[i] < prefixList[i]:
                trapped += min(suffixList[i], prefixList[i]) - height[i]
        return trapped

File: 42-trapping-rain-water/trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int[] prefixArray = new int[height.length + 1];
        int[] suffixArray = new int[height.length + 1];
        int n = height.length + 1;
        for (int i = 1; i < n; i++) {
            prefixArray[i] = Math.max(prefixArray[i - 1], height[i - 1]);
        }
        for (int i = n - 2; i >= 0; i--) {
            suffixArray[i] = Math.max(suffixArray[i + 1], height[i]);
        }
        // System.out.println(Arrays.toString(prefixArray) + " " + Arrays.toString(suffixArray)); 
        int maxArea = 0;
        for (int i = 0; i < height.length; i++ ){
            maxArea += Math.min(prefixArray[i+1], suffixArray[i]) - height[i];
        }
        return maxArea;
    }
}